File: src/main/java/io/github/saneea/feature/print/PrintLine.java
package io.github.saneea.feature.print;

import java.io.PrintStream;
import java.util.function.Consumer;

import org.apache.commons.cli.CommandLine;
import org.apache.commons.cli.Option;

import io.github.saneea.Feature;
import io.github.saneea.FeatureContext;

public class PrintLine implements //
		Feature, //
		Feature.CLI, //
		Feature.Out.Text.PrintStream //
{

	private static final String TEXT = "text";
	private static final String NO_NEW_LINE = "noNewLine";

	private PrintStream out;
	private CommandLine commandLine;

	@Override
	public String getShortDescription() {
		return "print text to output";
	}

	@Override
	public void run(FeatureContext context) throws Exception {

		String text = commandLine.getOptionValue(TEXT);

		Consumer<String> printFunc = //
				commandLine.hasOption(NO_NEW_LINE)//
						? out::print //
						: out::println;

		printFunc.accept(text);
	}

	@Override
	public void setOut(PrintStream out) {
		this.out = out;
	}

	@Override
	public Option[] getOptions() {
		return new Option[] { //

				Option//
						.builder("t")//
						.longOpt(TEXT)//
						.hasArg(true)//
						.argName("text for printing")//
						.required(true)//
						.desc("e.g. 'Hello World'")//
						.build(), //

				Option//
						.builder("nnl")//
						.longOpt(NO_NEW_LINE)//
						.required(false)//
						.desc("do not add new line at the end of output")//
						.build()//
		};
	}

	@Override
	public void setCommandLine(CommandLine commandLine) {
		this.commandLine = commandLine;
	}

}

File: src/main/java/io/github/saneea/dvh/feature/print/PrintLine.kt
package io.github.saneea.dvh.feature.print

import io.github.saneea.dvh.Feature
import io.github.saneea.dvh.Feature.Meta
import io.github.saneea.dvh.FeatureContext
import org.apache.commons.cli.CommandLine
import org.apache.commons.cli.Option
import java.io.PrintStream

class PrintLine :
    Feature,
    Feature.CLI,
    Feature.CLI.Options,
    Feature.Out.Text.PrintStream {

    override lateinit var context: FeatureContext
    override lateinit var outTextPrintStream: PrintStream
    override lateinit var commandLine: CommandLine

    override val meta = Meta("print text to output")

    override fun run() {
        val text = commandLine.getOptionValue(TEXT)

        val printFunc: (String) -> Unit =
            if (commandLine.hasOption(NO_NEW_LINE))
                outTextPrintStream::print
            else
                outTextPrintStream::println

        printFunc(text)
    }

    override val options
        get() = arrayOf(
            Option
                .builder("t")
                .longOpt(TEXT)
                .hasArg(true)
                .argName("text for printing")
                .required(true)
                .desc("e.g. 'Hello World'")
                .build(),
            Option
                .builder("nnl")
                .longOpt(NO_NEW_LINE)
                .required(false)
                .desc("do not add new line at the end of output")
                .build()
        )

    companion object {
        private const val TEXT = "text"
        private const val NO_NEW_LINE = "noNewLine"
    }
}